Optimizes Runechat

Kylerace opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
requested by @Mothblocks 
during the winter ball we disabled runechat for most of the event and had decent time dilation numbers - and then we enabled runechat and within minutes it was one of the most expensive things on the server and increased time dilation by a pretty large amount. so i set out to make runechat scalable to events with 100+ population because not having it is annoying and takes away from events where its disabled. 

the way that runechat works on master is that when something speaks and wants to relay it to n client mobs that heard the message, it creates n /datum/chatmessage instances which handle creating the runechat image, setting up the first animation stage (going from 0 alpha to 255 over a short duration) and adding a timer for the fading stage (going from 255 alpha to 0 over a short duration), and also editing the timers for all other messages already said by the speaker and heard by the current hearer. all of this is pretty inefficient, ideally if byond had reasonable clientside support we would just need to send n clients the information relevant to constructing the runechat image itself but we dont have that. 

now images from the same message are handled by a single chatmessage datum, each client still gets its own image which was one of the goals of setting out to do this but turned out to be really hard and honestly not worth it. each image is then given its entire animation sequence at once, it no longer needs to have its fade manually edited by the server after a set period of time its all done by the client. now each message sent only needs 1 timer insertion total, regardless of how many hearers there are, and editing older messages no longer requires another timer insertion though it does require some work in sending a new animation sequence. 
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->
## What Could Still Be Done Better
there are at least two more big things that could be done but both would be ass to work through. currently the cost of creating a new message scales with the number of older messages already seen from the speaker to the hearer because it has to iterate all of the still existing messages and animate them upwards and make them fade away faster (so they dont take up the screen). i could create a "column" image for each client and attach the maptext to that for each new message and animate it upwards. that would make the cost of creating a new image not scale with the number of old images however i would need to find a way to mask the column such that older messages fade away faster than they otherwise would have and im not sure how i would do that. 

another option is somehow replacing the client/MeasureText() call with either a good approximation or a perfect solution thats synchronous and fast. currently creating an image requires the client to receive the call to send the dimensions of the text and to send it back to the server, across n clients that adds up and itd be nice to make that synchronous somehow. @LemonInTheDark told me they think they could do it, so we can see if its needed once we stress test this live. 

## Why It's Good For The Game
better scalability. having runechat (hopefully) enabled for events is nice.
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Kylerace committed:
makes runechat only create one chatmessage per message to save work

Kylerace committed:
docs the structure of /client/var/list/seen_messages

Kylerace committed:
fixes dumb bug with mob/abstract_move

Kylerace committed:
makes SSrunechat into a custom timing style SS to guarentee O(1) insert

Kylerace committed:
helps make chatmessage work with the new version of SSrunechat

Kylerace committed:
removes timer steps - makes runechat messages do all steps via animating

Kylerace committed:
makes timers_by_hash work right

Kylerace committed:
makes the image fading animations look better

Kylerace committed:
makes SSrunechat a timer subsystem again because it already does what i wanted to do lol

Kylerace committed:
removes unneeded shit from chatmessage datums and finally fixes the client animations

Kylerace committed:
deletes benchmark code (for now *evil laugh*)

Kylerace committed:
documentation addition and fix

Kylerace committed:
fixes issue with messages being reset to 255 alpha if forced to fade

Kylerace committed:
saves some time by creating the static MA via a proc initializing consts

Kylerace committed:
do not look at the man behind the curtain

LemonInTheDark commented:
On the topic of MeasureText, if it is like I suspect just a matter of the width of the client's window, we can hook into that, so we'd only need to do one async call per message, and then use that as the basis for all others, rather then one per viewer. 
Would need to hook into resizing, but that saves headache in other places, and can just go on the view datum, so it's all good.

If it's not just map width we might be doomed.

LemonInTheDark commented on code/__HELPERS/_lists.dm:
```diff
@@ -51,6 +51,8 @@
 ///Returns the list if it's actually a valid list, otherwise will initialize it
 #define SANITIZE_LIST(L) ( islist(L) ? L : list() )
 #define reverseList(L) reverse_range(L.Copy())
+/// inserts the given Key Value pair as a list association in L at the given index.
+#define INSERT_ASSOCIATIVE_KV_PAIR(L, K, V, _index) L.Insert(_index, K); L[K] = V;
```
Fucking cursed brother. Is this even used or am I blind

Kylerace replied:
it was when i made SSrunechat into a non SStimer subtype :trollface:

LemonInTheDark replied:
Pogged

Kylerace replied:
im sure someone could use it

LemonInTheDark replied:
it'd break after the first insert, since one would just shift all the others over 
s not a good pattern I don't think

Kylerace replied:
not useful anyways, removed

Kylerace committed:
removes (almost) useless define i dont need anymore

Kylerace committed:
splits up generate_image() so that MeasureText doesnt cause overtime.

now once the client returns its text size the second proc is queued up and processed by the subsystem the next tick

Mothblocks commented:
@LemonInTheDark It's going to be based on the actual font itself more than anything, might even care about DPI and such.

We can probably reverse engineer it fairly easily to be honest.

Kylerace committed:
makes MeasureText() only get called once per client instead of twice

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into efficient_runechat

Kylerace commented:
if i figure out the rustg implementation of MeasureText() i can pr that later. in any case it no longer calls MeasureText() twice in a row per client and only does it once. this should be a lot more scalable than the original

Kylerace committed:
whoops, fixes runtime in on_hearer_qdel

Kylerace committed:
fixes a null client runtime in generate_image

Kylerace committed:
fixes runtime in handle_new_image_association

GolfinhoVoador commented:
fixed by removing it seems

JohnFulpWillard commented:
> fixed by removing it seems

finally

Kylerace committed:
im not a smart man martha

Kylerace committed:
whoops i wasnt putting the deletion timers into SSrunechat

Kylerace committed:
docs for /datum/chatmessage/proc/prepare_text()

Kylerace committed:
adds more runtime protection that hopefully fixes tm issues

AnturK commented on code/datums/chatmessage.dm:
```diff
@@ -66,51 +59,93 @@
-	RegisterSignal(owned_by, COMSIG_PARENT_QDELETING, .proc/on_parent_qdel)
+	// Register client who hears this message
+	var/client/owned_by = owner.client
+	RegisterSignal(owned_by, COMSIG_PARENT_QDELETING, .proc/on_hearer_qdel)
```
This runtimes on ghosts with the current TM

LemonInTheDark replied:
God signals are depressing

Kylerace replied:
should be fixed from [`c92148e` (#65791)](https://github.com/tgstation/tgstation/pull/65791/commits/c92148ea793fd952ca0e403cde5a6663d8a207b3)

Kylerace committed:
moves client registration to when its actually needed

also improves create_chat_message() a little bit

LemonInTheDark commented on code/__DEFINES/dcs/helpers.dm:
```diff
@@ -20,3 +20,9 @@
 /// A wrapper for _LoadComponent that allows us to pretend we're using normal named arguments
 #define LoadComponent(arguments...) _LoadComponent(list(##arguments))
+
+/atom/proc/dcs_usage()
+
+/proc/list_usage(list/input_list)
+
+/proc/is_assoc_element(list/input_list, index)
```
Brother what

Kylerace replied:
AH FUCK GOD DAMMIT THIS ALWAYS HAPPENS

Kylerace committed:
meninblackflash.gif

Kylerace committed:
adds client tracking before MeasureText(). should fix tm errors

Kylerace committed:
actually for real should fix tm hard deletes i hope oh god please work

Kylerace committed:
Update code/__DEFINES/runechat.dm

Co-authored-by: Seth Scherer <supernovaa41@gmx.com>

SuperNovaa41 commented on code/__DEFINES/runechat.dm:
```diff
@@ -0,0 +1,27 @@
+/// How long the chat message's spawn-in animation will occur for
+#define CHAT_MESSAGE_SPAWN_TIME (0.2 SECONDS)
+/// How long the chat message will exist prior to any exponential decay
+#define CHAT_MESSAGE_LIFESPAN 5 SECONDS
```
```suggestion
#define CHAT_MESSAGE_LIFESPAN (5 SECONDS)
```

SuperNovaa41 commented on code/__DEFINES/runechat.dm:
```diff
@@ -0,0 +1,27 @@
+/// How long the chat message will exist prior to any exponential decay
+#define CHAT_MESSAGE_LIFESPAN 5 SECONDS
+/// How long the chat message's end of life fading animation will occur for
+#define CHAT_MESSAGE_EOL_FADE 0.7 SECONDS
```
```suggestion
#define CHAT_MESSAGE_EOL_FADE (0.7 SECONDS)
```

SuperNovaa41 reviewed (commented):
i guess i should've suggested all of these too

Kylerace committed:
Update code/__DEFINES/runechat.dm

Co-authored-by: Seth Scherer <supernovaa41@gmx.com>

Kylerace committed:
Update code/__DEFINES/runechat.dm

Co-authored-by: Seth Scherer <supernovaa41@gmx.com>

LemonInTheDark commented on code/datums/chatmessage.dm:
```diff
@@ -168,69 +230,145 @@
-	addtimer(CALLBACK(GLOBAL_PROC, /proc/qdel, src), fadetime, TIMER_DELETE_ME, SSrunechat)
+	return message
+
+/proc/create_runechat_template()
+	var/mutable_appearance/template = new()
+	template.maptext_width = CHAT_MESSAGE_WIDTH
+	template.plane = RUNECHAT_PLANE
+	template.appearance_flags = APPEARANCE_UI_IGNORE_ALPHA | KEEP_APART
+	template.alpha = 0
+	template.maptext_width = CHAT_MESSAGE_WIDTH
```
could this just be a subtype? or is that too cursed

Kylerace replied:
i tried that and it didnt work for some reason, might be due to using an MA

LemonInTheDark commented on code/datums/chatmessage.dm:
```diff
@@ -255,11 +391,23 @@
+		text_to_use = lang_treat(speaker, message_language, raw_message, spans, null, TRUE)
+		spans = spans ? spans.Copy() : list()
+
+	message_to_use = SSrunechat.messages_by_creation_string["[text_to_use]-[REF(speaker)]-[message_language]-[list2params(spans)]-[CHAT_MESSAGE_LIFESPAN]-[world.time]"]
```
if you're going to generate a key here, and inside /datum/chatmessage, why not pass the key in to avoid needing to do it twice?

Kylerace replied:
good idea

Kylerace commented on code/controllers/subsystem/runechat.dm:
```diff
@@ -1,3 +1,18 @@
+	///queue for runechat messages receiving client MeasureText() calls and ready to create the runechat images
+	var/list/datum/callback/message_queue = list()
+
+/datum/controller/subsystem/timer/runechat/fire(resumed)
+	. = ..() //poggers
+	while(message_queue.len)
+		var/datum/callback/queued_message = message_queue[message_queue.len]
+		queued_message.Invoke()
+		message_queue.len--
+		if(MC_TICK_CHECK)
+			return
```
yes, but its guaranteed to go through at least one message because it checks tick after one

Kylerace commented on code/datums/chatmessage.dm:
```diff
@@ -168,69 +230,145 @@
+ */
+/datum/chatmessage/proc/generate_image(atom/target, mob/owner, complete_text, mheight)
+	var/client/owned_by = owner.client
+	if(!owned_by || QDELETED(target) || QDELETED(src))//possible now since generate_image() is called via a queue
```
fair point

Kylerace replied:
```suggestion
	if(!owned_by || QDELETED(target) || QDELETED(src))//possible since generate_image() is called via a queue
```

Kylerace committed:
Update code/datums/chatmessage.dm

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

Kylerace commented on code/datums/chatmessage.dm:
```diff
@@ -66,51 +63,116 @@
+
+	SSrunechat.messages_by_creation_string[creation_parameters] = src
+
+	messages = list()
+	hearers = list()
+	approx_lines = list()
+	fade_times_by_image = list()
+	all_hearers = list()
```
not one ive measured no

LemonInTheDark replied:
why not do this in init then? preventing early calls?

Kylerace commented on code/datums/chatmessage.dm:
```diff
@@ -1,59 +1,56 @@
 	/// The current index used for adjusting the layer of each sequential chat message such that recent messages will overlay older ones
 	var/static/current_z_idx = 0
-	/// Contains ID of assigned timer for end_of_life fading event
+	/// Contains the hash of our main assigned timer for the end_of_life fading event. by the time this timer executes all clients should have
```
its a hash isnt it?

LemonInTheDark replied:
I mean I guess

Kylerace committed:
removes template_message because its unused

Kylerace commented on code/datums/chatmessage.dm:
```diff
@@ -1,59 +1,56 @@
-	/// Contains the reference to the previous chatmessage in the bucket, used by runechat subsystem
-	var/datum/chatmessage/prev
+	/// the full edited message the speaker created, might be edited by subsequent hearers
+	var/template_message = ""
```
this isnt even used anymore, removed

Kylerace commented on code/datums/chatmessage.dm:
```diff
@@ -66,51 +63,116 @@
  */
-/datum/chatmessage/proc/generate_image(text, atom/target, mob/owner, datum/language/language, list/extra_classes, lifespan)
+/datum/chatmessage/proc/prepare_text(text, atom/target, mob/owner, datum/language/language, list/extra_classes)
+	set waitfor = FALSE //because this waits on client input
```
yes

Kylerace replied:
```suggestion
	set waitfor = FALSE //because this waits on info passed from the client
```

Kylerace committed:
Update code/datums/chatmessage.dm

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

Kylerace commented on code/datums/chatmessage.dm:
```diff
@@ -168,69 +230,145 @@
+	if(!owned_by || QDELETED(src) || QDELETED(message_loc) || !all_hearers?[owned_by])
+		return //we should already have been qdel'd() if this evaluates to TRUE, doing it now would throw an error
+
+	var/mheight = WXH_TO_HEIGHT(measurement)
+	if(TICK_CHECK)
+		SSrunechat.message_queue += CALLBACK(src, .proc/generate_image, target, owner, complete_text, mheight)
```
we need to use our height when moving older messages upwards

LemonInTheDark replied:
I mean more why are we calling WXH_TO_HEIGHT here and not inside generate_image

Kylerace replied:
the reason why generate_image() and prepare_text() are two different procs now is because of MeasureText(), the measurement is needed for further processing but reading it puts us into the verb stage of the tick when we're done. if it were in generate_image() we'd have done nothing other than introduce another proc call

Kylerace replied:
since we have SSspeech_controller, before we call MeasureText() we should be either 1. the verb stage of the tick BUT the server isnt overloaded, or 2. near the beginning of the tick since the server was overloaded when the client's say() input reached the server so it was queued for the next tick. 

if we dont queue after each client input-waiting proc in a line of processing, we will be putting processing in the potentially dangerous verb portion of the tick where the server may or may not be overloaded, and we're doing stuff at the end of the tick.

LemonInTheDark replied:
I'm not saying anything about measure text.
WXH_TO_HEIGHT does not yield, or do anything funky. it just splits some text, and hands back a number.
Why are we doing that splitting before the tick check, rather then inside generate_image.
Also what do you mean by reading it?

LemonInTheDark replied:
ah it got refactored while we were talking, nvm potentially, minor overhead but it's a text op and a text2num, not that bad

Kylerace commented on code/datums/chatmessage.dm:
```diff
@@ -1,59 +1,56 @@
-#define CHAT_LAYER_Z_STEP 0.0001
-/// The number of z-layer 'slices' usable by the chat message layering
-#define CHAT_LAYER_MAX_Z (CHAT_LAYER_MAX - CHAT_LAYER) / CHAT_LAYER_Z_STEP
+///the layer value given to the message image in the last animation step after the image has faded.
+///used for either the initialization animation sequence or the edited stage 2 duration animation sequence.
+///to the server the images layer is set instantly to this value, but the client cant see it because its only adjusted for them
+/// after the message is inivisible. so this is used to mark what animation is being used for the server.
+#define MESSAGE_ANIMATION_DEFAULT_LAYER_MARK 1020
+///the layer value given to the message image for the animation sequence where the image is forced into fading.
+///used so that the server doesnt force the image to fade twice.
+#define MESSAGE_ANIMATION_FORCE_FADE_LAYER_MARK 1021
+///the layer mark given to the last stage of the animation sequence after it has been edited but not forced into the fading animation.
+#define MESSAGE_ANIMATION_EDIT_FADE_LAYER_MARK 1022
```
no because its not real when you consider it affecting visuals that clients see. it never actually visually gets to that layer for the client because its only done after the message fade animations have already completed and the alpha becomes 0. to the server it does instantly set layer so we can use it to tell what animation we're in

Kylerace committed:
potato reviews

Kylerace committed:
Merge remote-tracking branch 'origin/efficient_runechat' into efficient_runechat

Kylerace committed:
idx -> num_seen_messages

Kylerace commented on code/datums/chatmessage.dm:
```diff
@@ -168,69 +230,145 @@
+			var/current_stage_2_time_left = preexisting_message.fade_times_by_image[other_message_image] - (world.time + CHAT_MESSAGE_SPAWN_TIME)
+
+			//how much time remains in the "fully visible" stage of animation, after we adjust it. round it down to the nearest tick
+			var/real_stage_2_time_left = round((current_stage_2_time_left) * (CHAT_MESSAGE_EXP_DECAY ** idx++) * (CHAT_MESSAGE_HEIGHT_DECAY ** combined_height), world.tick_lag)
```
done

Kylerace committed:
changes WXH_TO_HEIGHT() for lemon

Kylerace commented:
> Emoting with custom say emotes forces runechat to say "An interesting thing to say" when they are only given one argument. Example: `coughs violently into their fist!*` would show up in runechat "An interesting thing to say"

@thgvr i cant replicate this on live or on local

Kylerace committed:
SHOULD fix hard deletes if it doesnt im ref tracking on live

Kylerace committed:
Merge remote-tracking branch 'upstream/master' into efficient_runechat

Kylerace committed:
whoops fixes Destroy runtime

Kylerace committed:
more safety and removes all_hearers list, moves it into hearers

Kylerace committed:
enables reference tracking for when im online to find hard dels

LemonInTheDark commented:
Will bully kyler about reopening this when he comes back

Kylerace commented:
@LemonInTheDark probably. its not one of my priorities right now though considering how annoying it was to work with. i might just do the queue thing and leave it at that
